src/vm/ssa.nit

```diff
@@ -0,0 +1,1140 @@
+
+		# Add the self variable
+		if self.selfvariable != null then variables.add(selfvariable.as(not null))
+
+		# Recursively goes into the nodes
+		if n_block != null then
+			n_block.generate_basicBlocks(vm, basic_block.as(not null))
+			is_generated = true
+		end
+	end
+end
+
+redef class AMethPropdef
+
+	# The return variable of the propdef
+	# Create an empty variable for the return of the method
+	# and treat returns like variable assignments
+	var returnvar: Variable = new Variable("returnvar")
+
+	redef fun generate_basicBlocks(vm)
+	do
+		basic_block = new BasicBlock
+		basic_block.first = self
+		basic_block.last = self
+
+		# If the method has a signature
+		if n_signature != null then
+			for p in n_signature.n_params do
+				# Add parameters to the local variables
+				variables.add(p.variable.as(not null))
+			end
+		end
+
+		# Add the return variable
+		variables.add(returnvar)
+
+		# Add the self variable
+		if self.selfvariable != null then variables.add(selfvariable.as(not null))
+
+		# Recursively goes into the nodes
+		if n_block != null then
+			n_block.generate_basicBlocks(vm, basic_block.as(not null))
+			is_generated = true
+		end
+	end
+end
+
+# Utility class for dump basic block and SSA output to dot files
+class BlockDebug
+	# The output file
+	var file: FileWriter
+
+	# Dump all the hierarchy of BasicBlock from `block` to the leaves
+	fun dump(block: BasicBlock)
+	do
+		# Write the basic blocks hierarchy in output file
+		file.write("digraph basic_blocks\n\{\n")
+		var i = 0
+		file.write(print_block(block, i))
+		file.write("\n\}")
+
+		file.close
+	end
+
+	# Print all the block recursively from `block` to the leaves
+	# *`block` The root BasicBlock
+	# *`i` Used for the recursion
+	private fun print_block(block: BasicBlock, i: Int): String
+	do
+		# Precise the type and location of the begin and end of current block
+		var s = "block{block.hash.to_s} [shape=record, label="+"\"\{"
+		s += "block" + block.to_s.escape_to_dot
+		s += "|\{" + block.first.location.file.filename.to_s + block.first.location.line_start.to_s
+		s += " | " + block.first.to_s.escape_to_dot
+
+		# Print phi-functions if any
+		for phi in block.phi_functions do
+			s += " | " + phi.to_s.escape_to_dot + " "
+		end
+
+		s += "}|\{" + block.last.location.file.filename.to_s + block.last.location.line_start.to_s
+		s += " | " + block.last.to_s.escape_to_dot + "}}\"];"+ "\n"
+
+		i += 1
+		block.treated = true
+
+		for b in block.successors do
+			# Print edges to successors
+			s += "block{block.hash.to_s} -> " + " block{b.hash.to_s};\n"
+
+			# Recursively print child blocks
+			if not b.treated then s += print_block(b, i)
+		end
+
+		return s
+	end
+end
+
+redef class AExpr
+	# Generate recursively basic block for this expression
+	# *`vm` Running instance of the virtual machine
+	# *`block` A basic block not completely filled
+	# Return the last created block
+	fun generate_basicBlocks(vm: VirtualMachine, block: BasicBlock): BasicBlock
```

Comment:
You should try to keep your parameter names consistent in redefinitions; so `block` or `old_block`?